common/digest/sha1.c

```diff
@@ -0,0 +1,230 @@
+    md->state[0] = md->state[0] + a;
+    md->state[1] = md->state[1] + b;
+    md->state[2] = md->state[2] + c;
+    md->state[3] = md->state[3] + d;
+    md->state[4] = md->state[4] + e;
+
+    return 0;
+}
+
+static int sha1_init(struct sha1_state * md)
+{
+   if (md == NULL) return -1;
+   md->state[0] = 0x67452301UL;
+   md->state[1] = 0xefcdab89UL;
+   md->state[2] = 0x98badcfeUL;
+   md->state[3] = 0x10325476UL;
+   md->state[4] = 0xc3d2e1f0UL;
+   md->curlen = 0;
+   md->length = 0;
+   return 0;
+}
+
+static int sha1_process(struct sha1_state * md, const unsigned char *in, unsigned long inlen)
+{
+    unsigned long n;
+    int err;
+    if (md == NULL) return -1;
+    if (in == NULL) return -1;
+    if (md->curlen > sizeof(md->buf)) {
+        return -1;
+    }
+    if (((md->length + inlen * 8) < md->length)
+        || ((inlen * 8) < inlen)) {
+        return -1;
+    }
+    while (inlen > 0) {
+        if (md->curlen == 0 && inlen >= 64) {
+            if ((err = s_sha1_compress(md, in)) != 0) {
+                return err;
+            }
+            md->length += 64 * 8;
+            in         += 64;
+            inlen      -= 64;
+        } else {
+            n = MIN(inlen, (64 - md->curlen));
+            memcpy(md->buf + md->curlen, in, (size_t)n);
+            md->curlen += n;
+            in         += n;
+            inlen      -= n;
+            if (md->curlen == 64) {
+                if ((err = s_sha1_compress(md, md->buf)) != 0) {
+                    return err;
+                }
+                md->length += 8 * 64;
+                md->curlen = 0;
+            }
+        }
+    }
+    return 0;
+}
+
+static int sha1_done(struct sha1_state * md, unsigned char *out)
+{
+    int i;
+
+    if (md == NULL) return -1;
+    if (out == NULL) return -1;
+
+    if (md->curlen >= sizeof(md->buf)) {
+       return -1;
+    }
+
+    /* increase the length of the message */
+    md->length += md->curlen * 8;
+
+    /* append the '1' bit */
+    md->buf[md->curlen++] = (unsigned char)0x80;
+
+    /* if the length is currently above 56 bytes we append zeros
+     * then compress.  Then we can fall back to padding zeros and length
+     * encoding like normal.
+     */
+    if (md->curlen > 56) {
+        while (md->curlen < 64) {
+            md->buf[md->curlen++] = (unsigned char)0;
+        }
+        s_sha1_compress(md, md->buf);
+        md->curlen = 0;
+    }
+
+    /* pad upto 56 bytes of zeroes */
+    while (md->curlen < 56) {
+        md->buf[md->curlen++] = (unsigned char)0;
+    }
+
+    /* store length */
+    STORE64H(md->length, md->buf+56);
+    s_sha1_compress(md, md->buf);
+
+    /* copy output */
+    for (i = 0; i < 5; i++) {
+        STORE32H(md->state[i], out+(4*i));
+    }
+
+    return 0;
+}
+
+void sha1(const void *in, unsigned long inlen, void* out)
+{
+    struct sha1_state ctx;
+    sha1_init(&ctx);
+    sha1_process(&ctx, (const unsigned char*)in, inlen);
+    sha1_done(&ctx, (unsigned char *)out);
```

Comment:
Not the case here, but I generally secure_memset the context after calculating the hash value. This prevents information leak on the stack. Perhaps better to change that here and in other hash implementations to avoid using the dangerous implementation where it matters.

Reply:
We don't need to do it here because we do not perform any security-sensitive operations. The original code from LibTomCrypt already has that resolved, but I explicitly removed that part because we do not care, and, in case of UEFITool-based tools will never start caring.